notify/jira/jira.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	switch issueSearchResult.Total {
+	case 0:
+		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
+		return nil, false, nil
+	default:
+		level.Warn(n.logger).Log("msg", "more than one issue matched, picking most recently resolved", "alert", key.String(), "picked", issueSearchResult.Issues[0].Key)
+
+		fallthrough
+	case 1:
+		return &issueSearchResult.Issues[0], false, nil
+	}
+}
+
+func (n *Notifier) getIssueTransitionByName(tmplTextFunc templateFunc, issueKey, transitionName string) (string, bool, error) {
+	path := fmt.Sprintf("issue/%s/transitions", issueKey)
+
+	responseBody, shouldRetry, err := n.doAPIRequest(tmplTextFunc, http.MethodGet, path, nil)
+	if err != nil {
+		return "", shouldRetry, err
+	}
+
+	var issueTransitions issueTransitions
+	err = json.Unmarshal(responseBody, &issueTransitions)
+	if err != nil {
+		return "", false, err
+	}
+
+	for _, issueTransition := range issueTransitions.Transitions {
+		if issueTransition.Name == transitionName {
+			return issueTransition.ID, false, nil
+		}
+	}
+
+	return "", false, fmt.Errorf("can't find transition %s for issue %s", transitionName, issueKey)
+}
+
+func (n *Notifier) TransitionIssue(tmplText templateFunc, key notify.Key, issueKey, transitionName string) (bool, error) {
+	transitionID, shouldRetry, err := n.getIssueTransitionByName(tmplText, issueKey, transitionName)
+	if err != nil {
+		return shouldRetry, err
+	}
+
+	requestBody := issue{}
+	requestBody.Transition = &idNameValue{ID: transitionID}
+
+	path := fmt.Sprintf("issue/%s/transitions", issueKey)
+
+	level.Debug(n.logger).Log("msg", "transitions jira issue", "alert", key.String(), "key", issueKey, "transition", transitionName)
+	_, shouldRetry, err = n.doAPIRequest(tmplText, http.MethodPost, path, requestBody)
+
+	return shouldRetry, err
+}
+
+func (n *Notifier) doAPIRequest(tmplTextFunc templateFunc, method, path string, requestBody any) ([]byte, bool, error) {
+	var body io.Reader
+	if requestBody != nil {
+		var buf bytes.Buffer
+		if err := json.NewEncoder(&buf).Encode(requestBody); err != nil {
+			return nil, false, err
+		}
+
+		body = &buf
+	}
+
+	url := n.conf.APIURL.JoinPath(path)
+	req, err := http.NewRequest(method, url.String(), body)
+	if err != nil {
+		return nil, false, err
+	}
+
+	var token string
+	if n.conf.APIToken != "" {
+		token, err = tmplTextFunc(string(n.conf.APIToken))
+		if err != nil {
+			return nil, false, fmt.Errorf("template error: %w", err)
+		}
+	} else {
+		content, err := os.ReadFile(n.conf.APITokenFile)
+		if err != nil {
+			return nil, false, fmt.Errorf("read key_file error: %w", err)
+		}
+		token, err = tmplTextFunc(string(content))
+		if err != nil {
+			return nil, false, fmt.Errorf("template error: %w", err)
+		}
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept-Language", "en")
+	req.Header.Set("X-Force-Accept-Language", "true")
+	req.SetBasicAuth(n.conf.APIUsername, strings.TrimSpace(token))
```

Comment:
It wasn't obvious to me that the token would be used for BasicAuth. Also it seems to be deprecated/discouraged from Atlassian documentation (https://developer.atlassian.com/server/jira/platform/basic-authentication/).

Reply:
The documentation you reference goes for Jira Server (on premise). 

The documentation for Cloud REST API does not deprecate the Auth Basic authentication flow: https://developer.atlassian.com/cloud/jira/platform/basic-auth-for-rest-apis/ and its recommend method for manual REST calls.

No words about deprecation here. Since Jira REST API is https exclusively, Auth Basic is sufficient enough.

Reply:
Right I mixed up things. Having said that the Server docs encourage people to use Personal Access Tokens:

![image](https://github.com/prometheus/alertmanager/assets/2587585/5e0a90b8-8dbb-4fee-a8c2-9eef8bba00bb)